hnn_core/parallel_backends.py

```diff
@@ -0,0 +1,296 @@
+"""Parallel backends"""
+
+# Authors: Blake Caldwell <blake_caldwell@brown.edu>
+#          Mainak Jas <mainakjas@gmail.com>
+
+import os
+import sys
+import multiprocessing
+import shlex
+import pickle
+import codecs
+from warnings import warn
+from subprocess import Popen, PIPE
+
+_BACKEND = None
+
+
+def _gather_trial_data(sim_data, net, n_trials):
+    """Arrange data by trial
+
+    To be called after simulate(). Returns list of Dipoles, one for each trial,
+    and saves spiking info in net (instance of Network).
+    """
+    dpls = []
+
+    for idx in range(n_trials):
+        dpls.append(sim_data[idx][0])
+        spikedata = sim_data[idx][1]
+        net.spikes._times.append(spikedata[0])
+        net.spikes._gids.append(spikedata[1])
+        net.gid_dict = spikedata[2]  # only have one gid_dict
+        net.spikes.update_types(net.gid_dict)
+
+    return dpls
+
+
+class JoblibBackend(object):
+    """The JoblibBackend class.
+
+    Parameters
+    ----------
+    n_jobs : int | None
+        The number of jobs to start in parallel. If None, then 1 trial will be
+        started without parallelism
+
+    Attributes
+    ----------
+    n_jobs : int
+        The number of jobs to start in parallel
+    """
+    def __init__(self, n_jobs=1):
+        self.type = 'joblib'
+
+        self.n_jobs = n_jobs
+        print("joblib will run over %d jobs" % (self.n_jobs))
+
+    def _parallel_func(self, func):
+        if self.n_jobs != 1:
+            try:
+                from joblib import Parallel, delayed
+            except ImportError:
+                warn('joblib not installed. Cannot run in parallel.')
+                self.n_jobs = 1
+        if self.n_jobs == 1:
+            my_func = func
+            parallel = list
+        else:
+            parallel = Parallel(self.n_jobs)
+            my_func = delayed(func)
+
+        return parallel, my_func
+
+    def __enter__(self):
+        global _BACKEND
+
+        self._old_backend = _BACKEND
+        _BACKEND = self
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        global _BACKEND
+
+        _BACKEND = self._old_backend
+
+    def _clone_and_simulate(self, net, trial_idx):
+        # avoid relative lookups after being forked by joblib
```

Comment:
why is that?

Reply:
`_clone_and_simulate` would fail. I just ran it again with a relative lookup:
```
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_joblib ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/blake/miniconda3/envs/mne/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 418, in _process_worker
    r = call_item()
  File "/Users/blake/miniconda3/envs/mne/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 272, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/Users/blake/miniconda3/envs/mne/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 567, in __call__
    return self.func(*args, **kwargs)
  File "/Users/blake/miniconda3/envs/mne/lib/python3.7/site-packages/joblib/parallel.py", line 225, in __call__
    for func, args, kwargs in self.items]
  File "/Users/blake/miniconda3/envs/mne/lib/python3.7/site-packages/joblib/parallel.py", line 225, in <listcomp>
    for func, args, kwargs in self.items]
  File "/Users/blake/repos/hnn-core-new/hnn_core/parallel_backends.py", line 88, in _clone_and_simulate
    from .neuron import NeuronNetwork, _simulate_single_trial
KeyError: "'__name__' not in globals"
"""
```

Reply:
okay fair enough :-)